addons/stock/models/stock_warehouse.py

```diff
@@ -166,12 +166,69 @@ def write(self, vals):
                     any(depend in vals for depend in depends):
                 warehouse._create_or_update_global_routes_rules()
 
+            if 'active' in vals:
+                picking_type_ids = self.env['stock.picking.type'].with_context(active_test=False).search([('warehouse_id', '=', warehouse.id)])
+                move_ids = self.env['stock.move'].search([
+                    ('picking_type_id', 'in', picking_type_ids.ids),
+                    ('state', 'not in', ('done', 'cancel')),
+                ])
+                if move_ids:
+                    raise UserError(_('You still have ongoing operations for picking\
+                        types %s in warehouse %s') %
+                        (', '.join(move_ids.mapped('picking_type_id.name')), warehouse.name))
+                else:
+                    picking_type_ids.write({'active': vals['active']})
+                location_ids = self.env['stock.location'].with_context(active_test=False).search([('location_id', 'child_of', warehouse.view_location_id.id)])
+                picking_type_using_locations = self.env['stock.picking.type'].search([
+                    ('default_location_src_id', 'in', location_ids.ids),
+                    ('default_location_dest_id', 'in', location_ids.ids),
+                    ('id', 'not in', picking_type_ids.ids),
+                    ('warehouse_id', '=', warehouse.id),
+                ])
+                if picking_type_using_locations:
+                    raise UserError(_('%s use default source or destination locations\
+                        from warehouse %s that will be archived.') %
+                        (', '.join(picking_type_using_locations.mapped('name')), warehouse.name))
+                warehouse.view_location_id.write({'active': vals['active']})
+
+                rule_ids = self.env['stock.rule'].with_context(active_test=False).search([('warehouse_id', '=', self.id)])
+                # Only modify route that apply on this warehouse.
+                route_ids = warehouse.route_ids.filtered(lambda r: len(r.warehouse_ids) == 1)
+                # Active/archive all warehouse components.
+                route_ids.filtered(lambda r: not r.warehouse_ids - warehouse).write({'active': vals['active']})
```

Comment:
this filtered is useless i think

Reply:
nope, it is to avoid removing routes linked to multiple warehouses.

Reply:
but you already filter on route having a single warehouse_ids in the previous line

Reply:
true